lib/OpenQA/Schema/Result/JobModules.pm

```diff
@@ -235,4 +235,35 @@ sub save_results {
     path($self->job->result_dir, 'details-' . $self->name . '.json')->spurt(encode_json($results));
 }
 
+# Collect text output into JSON. save_results() is called too early for this.
+sub finalize_results {
+    my ($self) = @_;
+
+    my $dir = $self->job->result_dir();
+    return unless $dir;
+    my $file = path($dir, "details-" . $self->name . ".json");
+    return unless -e $file;
+    my $results;
+
+    try {
+        $results = decode_json($file->slurp);
+    }
+    catch {
+        return;
+    };
+
+    $results = {details => $results} if ref($results) eq 'ARRAY';
+
+    for my $step (@{$results->{details}}) {
+        next unless $step->{text};
+        my $txtfile = path($dir, $step->{text});
+        next unless -e $txtfile;
+        $step->{text_data} = $txtfile->slurp;
+    }
+
+    my $tmpfile = $file->sibling($file->basename . '.tmp');
+    $tmpfile->spurt(encode_json($results));
+    rename $tmpfile->to_string, $file->to_string;
+}
```

Comment:
It seems a little bit dangerous to do so much file handling within code which is executed to serve an HTTP request. This code might run in parallel (in production we use "preforking"). Besides, delaying the HTTP response until all of this file handling is done is problematic as well (the file system on our production servers can be under heavy load and this code might therefore be slow). Maybe it is better to just start a Minion job here? We use this in openQA generally to start background tasks and it supports locking.

Reply:
That sounds like a good idea. Where can I find some documentation on minion handling?

Reply:
That's the Perl module we're using: https://metacpan.org/release/Minion https://metacpan.org/pod/Minion and https://metacpan.org/pod/Minion::Job are the most important packages.

Within openQA we create packages like `OpenQA::Task::Job::Limit` for tasks to run via Minion. This one could be `OpenQA::Task::Job::FinalizeModules` and would be called once the job is done. Of course it would be possible to make a task for each module as well, e.g. `OpenQA::Task::JobModule::Finalize`.

Regardless how you name it, the name *must* be added to the `OpenQA::WebAPI::Plugin::Gru::register_tasks` function.

Tasks can be invoked via `OpenQA::WebAPI::Plugin::Gru::enqueue` like in the helper `OpenQA::WebAPI::Plugin::Gru::enqueue_limit_assets`.

Maybe we should add some documentation at some point.